Runtime device detection

joamaki opened the pull request:
This implements experimental detection of new and deleted datapath devices at runtime behind `--runtime-device-detection` flag.

The existing route-based mechanism for selecting new candidate devices is used. The DELLINK netlink
messages are used for detecting removed devices.

On device change this updates `option.Config.Devices` and reconfigures the datapath. Additionally
NodePort entries in the lbmap are updated to include the new frontend IPs of the added devices.

A K8sT test case is included which creates vxlan interfaces on two nodes and verifies that the device
is detected and datapath reloaded by accessing a NodePort service over the vxlan interface.

IPsec and BandwidthManager changes are left for follow-up PR.

joamaki committed:
datapath: Move DeviceManager to the datapath package

datapath/linux package is a more natural location for the
device detection code now that runtime detection will
introduce dependencies to DeviceManager from other packages.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
datapath: Runtime detection of devices

Adds support to DeviceManager to listen for device changes
via the NEWROUTE and DELLINK netlink messages.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Provide up-to-date status on devices

Recompute the devices for agent status response as they can
now change with runtime device detection.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
test: Add k8sT test for testing runtime device detection

Adds a k8sT test that tests the runtime detection by adding
a vxlan tunnel between the two nodes and checking that NodePort
services can be accessed using the new device.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Reload datapath on device changes

Add the --runtime-device-detection flag.

Implement the function to reload the datapath when devices change.

Since frontend IPs are populated when a service is created, we also
add the IPs of new devices to the lbmaps to make NodePort services
accessible over these new IPs. This mirrors how the agent would behave
if these devices would have been specified with --devices.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Use user provided list of devices to filter new devices

If the user set "--devices", use it to filter out new devices
detected at runtime.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

borkmann reviewed (approved):
LGTM 🚀 , nice work! Do you have a follow-up list as GH issue with changes needed after the initial merge? I presume the device regex support for runtime changes, ipsec/bwm and one thing that would be needed for policy is to push the new dev addr into the ipcache as host id (something related I recently stumbled into: https://github.com/cilium/cilium/pull/19513), also various tests for these features (incl. hostfw) in CI, right?

tklauser commented on daemon/cmd/daemon.go:
```diff
@@ -824,14 +824,6 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
-		// daemon options.
-		d.k8sCachesSynced = d.k8sWatcher.InitK8sSubsystem(d.ctx)
-		bootstrapStats.k8sInit.End(true)
-	}
```
Looks like the this hunk is added in commit `datapath: Runtime detection of devices` and then removed in the successive commit `daemon: Reload datapath on device changes` again. The same hunk is already present a few lines down. Is this addition/removal intentional?

tklauser reviewed (approved):
Nice!

One nit about some code being added and removed again in two successive commits.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -810,18 +810,28 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+		d.k8sCachesSynced = d.k8sWatcher.InitK8sSubsystem(d.ctx)
+		bootstrapStats.k8sInit.End(true)
+	}
+
```
The second patch adds these lines, which immediately sent :rotating_light: since we've had deadlock issues from rearranging k8s init in the past. However a later patch removes these lines again. To protect bisectability, would you mind removing them?

joamaki replied:
Weird, this must've been an artifact from rebasing. I'll clean this up.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -172,6 +173,15 @@ func (dm *DeviceManager) getDevices() []string {
 
+// TODO(JM): Try and get rid of this public method. Preferably device manager just provides
+// events on device change and if a component needs to know the current list of devices it
+// would maintain it itself. This to enforce that all components subscribe to the events.
```
Should be pretty easy, the only user seems to be tests and the one new user that you introduced in this PR:

```
$ git grep GetDevices
pkg/datapath/linux/devices.go:func (dm *DeviceManager) GetDevices() []string {
pkg/datapath/linux/devices.go:                          devices := dm.GetDevices()
pkg/datapath/linux/devices_test.go:             c.Assert(dm.GetDevices(), checker.DeepEquals, devices)
pkg/datapath/linux/devices_test.go:             c.Assert(dm.GetDevices(), checker.DeepEquals, devices)
pkg/datapath/linux/devices_test.go:             c.Assert(dm.GetDevices(), checker.DeepEquals, devices)
```

joamaki replied:
I initially renamed `option.Config.Devices` to `InitialDevices` and plan was to make everything use `GetDevices()`, hence the comment. I think I would at this point prefer instead to leave other code as is and then go refactor all uses of `option.Config.Devices` to be event driven instead. 
Removed the TODO.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+
 // expandDevices expands all wildcard device names to concrete devices.
 // e.g. device "eth+" expands to "eth0,eth1" etc. Non-matching wildcards are ignored.
+// FIXME(JM): Return the expanded devices, don't touch InitialDevices
```
What's the implication of a `FIXME` like this?

joamaki replied:
Oops, that's not relevant anymore. Removed.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+	linkChan := make(chan netlink.LinkUpdate)
+	err = netlink.LinkSubscribeWithOptions(linkChan, closeChan, netlink.LinkSubscribeOptions{
+		Namespace: netNS,
+		// FIXME(JM): What if a device is removed between Detect() and Listen()?
```
Could we write a unit test for this case? Seems like the kind of thing where you get the "once in a blue moon" report that something fails to load programs on a device, then if you restart the agent it's fine. Possible source of rare CI flakes (though I would hope we can catch & highlight the relevant error messages if that's the case).

joamaki replied:
I fixed the race and added a test.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+			// List existing routes to make sure we did not miss changes that happened after Detect().
+			// FIXME(JM): Perhaps could consider combining Detect() and Listen() so we don't have to do this.
+			// Having this whole thing be fully event-driven is likely a better design as that forces consumers
+			// to work with it correctly.
```
I like the idea of combining `Detect()` and `Listen()`. It seems to me like the desired behaviour is that from the current `Detect()` logic point, it should be able to read all of the existing device names off the channel, then once the initial listing is complete, hand the channel to a goroutine to deal with subsequent updates, and continue with the rest of the initialization. That doesn't seem _so_ far from what you already have today, except rather than starting to handle the events at the end of the daemon initialization, the goroutine starts earlier. 

Not sure if there's any tricky problems to solve with that? We'd probably want to ensure that this device update logic only happens after the initial datapath regeneration has occurred, I'm not sure if the `Trigger...` statement in `Daemon.ReloadOnDeviceChange()` already deals with this timing constraint.

At the same time, specifically for this concern, I think that we can first merge this PR and then continue to iterate on top to improve on this & simplify. I don't think that this necessarily has any direct implications today; the current implementation is fine, we're just considering if there's even better ways to do this in future.

joamaki replied:
The tricky problems are mostly with refactoring the downstream to work with device change events. I'd like to postpone that until the agent modularization as it'll slot in nicer at that point.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+		defer ticker.Stop()
+		buffer := make([]netlink.Route, 0)
+
+		log.Printf("Listening for device changes")
```
Did you mean `log.Info(...)`? I don't think we intend to use `log.Printf()` anywhere in the codebase. I don't even know what level that shows up as.

joamaki replied:
Thanks, fixed. I think at the time I started working on this I was only really familiar with Go's log package :)

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+			if err := netns.Set(*netNS); err != nil {
+				log.WithError(err).Warnf("Failed to set network namespace")
```
Log messages like this should not use `Warnf()`, `Warnf(...)` expects a formatting string. If there is no formatting string, it can use `Warn()`.

(I didn't check the whole PR for this, please double-check all the logs).

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+		// threads namespace. This is currently only relevant for testing.
+		if netNS != nil {
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
```
Mildly interesting point, I guess it probably doesn't matter here given the usage for tests, but this `defer()` will only be run when the goroutine exits, which would only happen once the context is cancelled.. not sure I understand the implications here given that the runtime needs to ensure that at least one thread is available in this namespace the whole time, so that this goroutine can execute in that namespace.

I've seen this pattern before with syscalls, but not so much to bound long-living goroutines.

(No action necessary, just pondering whether this is a pattern we should be vigilant about if it ever gets introduced outside test code).

joamaki replied:
This has been now removed. I instead opted for using `netlink.NewHandleAt` for the netlink operations so no need for the thread pinning.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+		// To avoid multiple reloads of the datapath when lots of routes are added in one go,
+		// collect the route updates into a batch and process the batch periodically.
+		//
+		ticker := time.NewTicker(dm.routeBatchingInterval)
```
I'm not sure if this is the right place to mitigate this concern.

I presume that `updateDevicesFromRoutes()` should relatively quickly identify whether a route introduces a new device or not, thereby filtering down the events such that we shouldn't signal to the caller too frequently, only if there truly are new devices showing up.

If the _caller_ needs to react to many devices changing at once, and ensure that it doesn't trigger the datapath reload too frequently beacuse _that_ code is expensive, then that's the concern of the caller. Hence the caller should have the code to mitigate this concern, probably via a `pkg/trigger`. I'm actually mildly surprised that `TriggerReloadWithoutCompile()` directly reinitializes the datapath without a trigger/ratelimiter, I had previously figured it was supposed to be async & ratelimited.

This way, we can simplify this core code around handling the channels and the interaction between the different events that the goroutine listens for, we can avoid a constant tick, and the ratelimiting is tied directly to the logic that's expensive.

joestringer replied:
(I note also that this ratelimiting is not extended to device deletes, so if there were many of those, then this code doesn't currently mitigate such a case).

joamaki replied:
I wasn't really thinking about the downstream logic, but more about reducing the amount of work done here if a lot of routes are added when a device comes up. However now that I think about it, this feels prematurely optimized. I'll drop it.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
 
 		viable := dm.isViableDevice(l3DevOK, info.hasDefaultRoute, link)
 		if viable {
+			log.WithField(logfields.Device, name).Info("Added new device")
```
I think that in an ideal world, we would probably log this _after_ reacting to the new device rather than before, so that the reader in the logs could say "Cilium now knows about this device and has reacted to it". I guess we could make this clearer, eg by saying "New device detected, evaluating which functionality to apply" or similar. This way if we end up debugging some issue that's related to device detection, then we will know how to interpret the log compared to when the functionality is successfully configured on this new device.

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
+
+			if devicesChanged {
+				devices := dm.GetDevices()
+				log.WithField(logfields.Devices, devices).Info("Detected new devices")
```
We will already log that we detected the individual device (previous comment). Do we need both?

Also, this can be triggered from link deletion, so the log message is not exactly accurate here.

joamaki replied:
Removed the comment for the individual device and left this one here. Changed it to "Detected devices".

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -141,23 +150,15 @@ func (dm *DeviceManager) Detect() error {
-	defer dm.Unlock()
-	return dm.getDevices()
+	deviceList := dm.getDevices()
+	log.WithField(logfields.Devices, deviceList).Info("Detected devices")
```
Hmm, now I look around at the logs it seems like we're sprouting multiple of this one. Maybe remove this one?

joamaki replied:
I'd keep both this one in `Detect()` and the one in `Listen()`, but make sure there's no other duplication.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
 
+// ReloadOnDeviceChange regenerates device related information and reloads the datapath.
+func (d *Daemon) ReloadOnDeviceChange(devices []string) {
+	log.WithField(logfields.Devices, devices).Infof("Datapath reload initiated due to changed devices.")
```
nit: Infof -> Info

Also wondering if this is crucial information to deliver at Info level or not. In `TriggerReloadWithoutCompile()` for instance we have a similar log at the beginning, but we leave that one at Debug level. Given the previous log messages that I commented about, it should already be clear that device changes have been detected, so there is a sign that Cilium is reacting to these changes. Unless we have significant new information to report here (such as success), I'm not sure that we need to log this. I believe that `TriggerReloadWithoutCompile()` will already cause the appropriate logs to be printed out once this regeneration is complete anyway, so maybe we can either drop this to debug or drop the line entirely?

joamaki replied:
Yep I think the "Devices changed" and the logs from datapath reload are enough. Dropping this line.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
+func (d *Daemon) ReloadOnDeviceChange(devices []string) {
+	log.WithField(logfields.Devices, devices).Infof("Datapath reload initiated due to changed devices.")
+
+	option.Config.Devices = devices
```
This makes me a little bit nervous, just because my understanding of the option package today is that there is zero locking protections, we only ever write it on startup, then reuse it all over the place. This is introducing a write operation arbitrarily later based on device changes.

joamaki replied:
Yep, I really want to get rid of this, either by switching everything to use `deviceManager.GetDevices()`, or preferably making the whole thing event based. Would you be OK with this for now, with this feature being behind an experimental flag and a having a follow up soon that does away with this? It's a bigger refactoring and I'd like to do it interleaved with the agent modularization as that makes the whole thing easier.

joestringer replied:
I agree that's the path forward.

Any chance there's a reasonable short-term fix with an atomic pointer or something?

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
+
+	if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade {
+		if err := node.InitBPFMasqueradeAddrs(devices); err != nil {
+			log.Warnf("InitBPFMasqueradeAddrs failed: %s", err)
```
log.WithError(err).Warn(...)

joestringer commented on pkg/service/service.go:
```diff
@@ -1401,3 +1405,74 @@ func isWildcardAddr(frontend lb.L3n4AddrID) bool {
+
+					_, _, err := s.upsertService(svc)
+					if err != nil {
+						log.WithError(err).Warnf("Could not create service for frontend %s", ip)
```
Please change this to use structured logging, `log.WithFields(logfields.IPAddr, ip).WithError(err).Warn(...)`. This allows logging systems to aggregate messages based on the message field and correlate the same type of warning between nodes etc. because the dynamic content (IP) is held in a different IP. 

(may or may not apply to other logging statements elsewhere, I won't go through all of them).

joamaki replied:
Done.

joestringer commented on pkg/service/service.go:
```diff
@@ -1401,3 +1405,74 @@ func isWildcardAddr(frontend lb.L3n4AddrID) bool {
+
+// SyncServicesOnDeviceChange finds and adds missing load-balancing entries for
+// new devices.
+func (s *Service) SyncServicesOnDeviceChange(nodeAddressing types.NodeAddressing) {
```
I didn't look too closely at this function, mostly I was focused on the daemon structure & how the datapath reinitialization pieces fit together (plus you have to set a timelimit somewhere). I presume someone else looked here already, but if you do want closer review then we can find someone to take a closer look or I can look again later.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1112,6 +1104,23 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+				}
+			}()
+		} else {
+			log.Warn("Runtime device detection requested, but no feature requires it. Disabling detection.")
```
Why `Warn`? This looks like it should be `Info`. If the user configured the particular set of options, then it's nothing to warn about, this is an OK state for the daemon to be in. We let the user know in case it surprises them, but it should not cause any unexpected consequences.

joamaki replied:
Changed to info.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
+	}
+
+	// Recreate node_config.h to reflect the mac addresses of the new devices.
+	if err := d.createNodeConfigHeaderfile(); err != nil {
```
What guarantees do we have about atomicity here?

Endpoints could be currently regenerating, plausibly reading this file at the same time as we do the rewrite..?

joamaki replied:
Good catch. The `InitBPFMasqueradeAddrs` and `InitNodePortAddrs` are also unsafe since they mutate maps that aren't protected as the assumption in the agent is that they're static... and probably `syncEndpointsAndHostIPs` is also unsafe too since that could be triggered from a controller concurrently with this.

I can fix the node config header issues by taking the compilation lock. The other issues related to mutating  globals (option.Config.Devices, node.ipv4NodePortAddrs, etc.) will require bigger refactoring. Feels a bit like all things like this that may change over time would need to be event-based to make sure there's reaction to the changes.

What do you think, should I start looking into making these mutations safer or would we be ok with putting this in behind an experimental flag and then fixing the potential races in a follow up?

joestringer replied:
I guess the 'simple' answer is to introduce some exclusion locks in the various places so that we at least protect these critical sections. Whether extending the critical section of the compilation lock, or adding individual locks for portions of this reconfiguration, or atomics. If there's a low-effort answer with these, then I think it would be worthwhile. Without that, I will have trouble suggesting anyone to turn the feature on. 

Out of curiousity, do we have a way to manually trigger the go race checker? Does it identify issues at this point? I believe we have a CI job for this which doesn't run on every PR, so it'd be useful to know whether this PR would introduce issues there; once you make a job like that break, it's difficult to get it back green again. Especially if we're assuming that more invasive work is required to address the issue.

joestringer replied:
yes, `/test-race`: https://docs.cilium.io/en/stable/contributing/testing/ci/#testing-with-race-condition-detection-enabled

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
 }
 
+// ReloadOnDeviceChange regenerates device related information and reloads the datapath.
+func (d *Daemon) ReloadOnDeviceChange(devices []string) {
```
Please document whether `devices` here represents the new devices, the deleted devices, or the complete set of devices that now exists.

joamaki replied:
Documented.

joestringer reviewed (changes requested):
Nice work. I've raised a few concerns here and there around critical sections / locking, just because unfortunately the new code could now start modifying state that was previously pretty static, and I'm not sure the full extent of the implications of that. I also commented on one or two areas where I think that the code be simplified.

Most of my feedback is actually just nits around the logging, probably one of these days I should just write up a linter so that we can enforce these consistently without me picking up on them during review. Some do require context, but some are general rules we tend to follow in the codebase that could be linted.

I reviewed most of the first four patches or so, and glanced at the rest. The combined presentation of unit tests + the code makes me reasonably confident that the filter should do the right thing anyway, and I figured that others have probably already reviewed that code in depth.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -1148,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
+	}
+
+	// Reload the datapath.
+	wg, err := d.TriggerReloadWithoutCompile("devices changed")
```
One more: Could this cause two goroutines to run the `Reinitialize()` of the datapath concurrently? Is that a problem?

joestringer replied:
OK, quick answer: No, it has a lock. But you'd do two back-to-back reinitalizations even if the first one successfully handles all the new devices. :shrug: I guess that's fine.

Feels a bit like we need an eventqueue here too, but I think that's more of an optimization rather than a necessary step to support this new functionality. I think we can get away without any further changes to this specific part for now.

joamaki committed:
datapath: Fix race with a deleted device after detection

If a device was removed after device detection, but before DeviceManager.Listen()
was invoked, then this deletion would be missed and the device would be kept.

Fix the issue by subscribing to link updates and then checking if some links
no longer exist.

While fixing this I noticed that a better way to make sure DeviceManager uses
the right namespace is to use the netlink.Handle instead of pinning the OS thread
and setting the namespace for the thread.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
pkg/node: Add lock to addresses

The runtime device reconfiguration will now update the addresses
for the new and deleted devices and thus the access to these must
be protected.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Protect option.Config.Devices with a mutex

option.Config.Devices is now changed when devices change, so it will
need to be protected from concurrent access. The ConfigPatchMutex cannot
be used as it's held when doing work that needs to access the devices
(OverrideEndpointOpts), so this adds 'devicesMu' and the GetDevices and
 SetDevices accessors.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -418,6 +418,14 @@ daemon:
 
+# -- Enables experimental support for the detection of new and removed datapath
+# devices. On device change the eBPF datapath, IPsec and BandwidthManager
+# functionality is reloaded to add and remove the device specific eBPF programs.
```
```suggestion
# functionalities are reloaded to add and remove the device-specific eBPF programs.
```

joamaki replied:
Thanks. Rewrote the comment completely since the IPsec and BandwidthManager support is TBD, and the wildcards are actually used now to filter the devices.

qmonnet reviewed (approved):
Nice work!

joestringer reviewed (approved):
LGTM, thanks!

joamaki committed:
daemon: Only run syncEndpointsAndHostIPs via a controller

ReloadOnDeviceChange called into syncEndpointsAndHostIPs directly and
this might race with the call from controller (runs once a minute).
Refactor the code so that we only call this function via the controller.

Since UpdateController immediately runs the function there is no need to
call syncEndpointsAndHostIPs directly before it, though we do want to check
if the first sync fails.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb commented on pkg/service/service.go:
```diff
@@ -1554,3 +1558,76 @@ func segregateBackends(backends []lb.Backend) (activeBackends map[string]lb.Back
+
+// SyncServicesOnDeviceChange finds and adds missing load-balancing entries for
+// new devices.
+func (s *Service) SyncServicesOnDeviceChange(nodeAddressing types.NodeAddressing) {
```
Damn, if we could get rid of this hack https://github.com/cilium/cilium/blob/master/pkg/k8s/service.go#L178, then this function would be way simpler :-/

joamaki replied:
What would it require to get rid of it?

brb replied:
Teaching pkg/service that NodePort service is provisioned with two frontends (0.0.0.0 and NodeIP). I think it's too complicated for this PR though.

joamaki committed:
service: Add sync test and impl. service removal

Add a test case for SyncServicesOnDeviceChange and implement
removal of services when a frontend IP has disappeared from
the load-balancer addresses.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb reviewed (approved):
:rocket: :moon:

sayboras commented on Documentation/spelling_wordlist.txt:
```diff
@@ -356,6 +356,8 @@ enableIPv
 enableIdentityMark
 enableIngressController
 enableK
+enableSecretsSync
```
i think this was added during rebase, however, master is not using this flag anymore, this can be removed.

sayboras reviewed (approved):
Looks good on helm + k8s related changes.
